UM/Resources.py

```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
+        cache_path = None
+        if platform.system() == "Windows":
+            cache_path = os.getenv("LOCALAPPDATA")
+        elif platform.system() == "Darwin":
+            cache_path = None
+        elif platform.system() == "Linux":
             try:
-                xdg_cache_home = os.environ["XDG_CACHE_HOME"]
+                cache_path = os.environ["XDG_CACHE_HOME"]
             except KeyError:
-                xdg_cache_home = os.path.expanduser("~/.cache")
-            cls.__cache_storage_path = os.path.join(xdg_cache_home, storage_dir_name)
-        else:
-            cls.__config_storage_path = "."
+                cache_path = os.path.expanduser("~/.cache")
 
-        if not cls.__data_storage_path:
-            cls.__data_storage_path = cls.__config_storage_path
+        return cache_path
 
-        if not cls.__cache_storage_path:
+    @classmethod
+    def __initializeStoragePaths(cls):
+        # use nested structure: <app-name>/<version>/...
+        if cls.ApplicationVersion == "master" or cls.ApplicationVersion == "unknown":
+            storage_dir_name = os.path.join(cls.ApplicationIdentifier, cls.ApplicationVersion)
+        else:
+            from UM.Version import Version
+            version = Version(cls.ApplicationVersion)
+            storage_dir_name = os.path.join(cls.ApplicationIdentifier, "%s.%s" % (version.getMajor(), version.getMinor()))
+
+        # config is saved in "<CONFIG_ROOT>/<storage_dir_name>"
+        cls.__config_storage_path = os.path.join(Resources.getConfigStorageRootPath(), storage_dir_name)
+
+        # data is saved in
+        #  - on Linux: "<DATA_ROOT>/<storage_dir_name>"
+        #  - on other: "<CONFIG_DIR>" (in the config directory)
+        data_root_path = Resources.getDataStorageRootPath()
+        cls.__data_storage_path = cls.__config_storage_path if data_root_path is None else \
+            os.path.join(data_root_path, storage_dir_name)
+
+        # cache is saved in
+        #  - on Linux:   "<CACHE_DIR>/<storage_dir_name>"
+        #  - on Windows: "<CACHE_DIR>/<storage_dir_name>/cache"
+        #  - on Mac:     "<CONFIG_DIR>/cache" (in the config directory)
+        cache_root_path = Resources.getCacheStorageRootPath()
+        if cache_root_path is None:
             cls.__cache_storage_path = os.path.join(cls.__config_storage_path, "cache")
+        else:
+            cls.__cache_storage_path = os.path.join(cache_root_path, storage_dir_name)
+            if platform.system() == "Windows":
+                cls.__cache_storage_path = os.path.join(cls.__cache_storage_path, "cache")
+
+        if not os.path.exists(cls.__config_storage_path):
+            cls._copyLatestDirsIfPresent()
 
         cls.__paths.insert(0, cls.__data_storage_path)
 
+    ##  Copies the directories of the latest version on this machine if present, so the upgrade will use the copies
+    #   as the base for upgrade
+    @classmethod
+    def _copyLatestDirsIfPresent(cls):
```

Comment:
The VersionUpgrader is an plugin, which means the plugins will already he loaded when VersionUpgrader is launched. A user can have custom plugins in his/her old app config directory, e.g. `cura/plugins`. If the VersionUpgrader decides to upgrade from that old version, Cura needs to restart to load those custom plugins. That's why we decide to run those code before actually doing anything.